db.js

```diff
@@ -426,6 +429,53 @@ exports.init = function (sbot, config) {
     )
   }
 
+  function addButtwoo(buffer, cb) {
+    const guard = guardAgainstDuplicateLogs('addButtwoo()')
+    if (guard) return cb(guard)
+
+    onceWhen(
+      stateFeedsReady,
+      (ready) => ready === true,
+      () => {
+        const data = butt2.extractData(buffer)
+        if (data.length !== 2)
+          return cb('incorrect buttwoo format, addButtwoo() failed')
+        if (data[0].length !== 3 || data[1].length !== 8)
+          return cb('incorrect buttwoo format, addButtwoo() failed')
+
+        const msgKeyBFE = butt2.hash(data)
+        const msgBIPF = butt2.butt2ToBipf(data, msgKeyBFE)
+
+        const author = bfe.decode(data[1][0])
+        const parent = bfe.decode(data[1][1])
+        const authorParent = author + (parent === null ? '' : parent)
+
+        const previous = state[authorParent] || { value: null }
+        const err = butt2.validateSingle(
+          data,
+          previous.value,
+          previous.key,
+          hmacKey
+        )
+
+        if (err)
+          return cb(clarify(err, 'message validation in addButtwoo() failed'))
+
+        state[authorParent] = {
+          key: msgKeyBFE,
+          value: data,
```

Comment:
@arj03 One thing I bumped into, and maybe this is new info for you: here we are setting the in-memory `state` to `authorParent => [encodedValue, sig, encodedContent]`, because `data` is that array with 3 parts.

However, when loading ssb-db2 from a cold start, the "latest" from Base index is loaded with KVTs (JS objects) from the log. So it's `author => KVT` that we'll have in `state`. Different shape for the state dictionary.

This means that after a cold start, if you call `addButtwoo`, the line 456 has a JS KVT (In fact, it's always going to have `null` as the previous.value, because the `state` is loaded with `author` from cold start, not `authorParent`). But subsequent `addButtwoo` calls will have an array with 3 parts on line 456.

In other words, the shape of `state` is not consistent here.

Reply:
Ah, we need a test for that. Good spotting.